numpy/core/src/umath/ufunc_object.c

```diff
@@ -646,253 +726,206 @@ get_ufunc_arguments(PyUFuncObject *ufunc,
-                    break;
-                case 'c':
-                    /* Provides a policy for allowed casting */
-                    if (strcmp(str, "casting") == 0) {
-                        if (!PyArray_CastingConverter(value, out_casting)) {
-                            goto fail;
-                        }
-                        bad_arg = 0;
-                    }
-                    break;
-                case 'd':
-                    /* Another way to specify 'sig' */
-                    if (strcmp(str, "dtype") == 0) {
-                        /* Allow this parameter to be None */
-                        PyArray_Descr *dtype;
-                        if (!PyArray_DescrConverter2(value, &dtype)) {
-                            goto fail;
-                        }
-                        if (dtype != NULL) {
-                            if (*out_typetup != NULL) {
-                                PyErr_SetString(PyExc_RuntimeError,
-                                    "cannot specify both 'signature' and 'dtype'");
-                                goto fail;
-                            }
-                            *out_typetup = Py_BuildValue("(N)", dtype);
-                        }
-                        bad_arg = 0;
-                    }
-                    break;
-                case 'e':
-                    /*
-                     * Overrides the global parameters buffer size,
-                     * error mask, and error object
-                     */
-                    if (strcmp(str, "extobj") == 0) {
-                        Py_INCREF(value);
-                        *out_extobj = value;
-                        bad_arg = 0;
+            /*
+             * Did we find a key and is this keyword desired (i.e., its
+             * output is not NULL)? Is so, convert it.
+             */
+            output = outputs[kwarg - ufunc_kwargs];
+            if (output) {
+                if (kwarg->convert) {
+                    if (!(*(kwarg->convert))(value, output)) {
+                        goto kwd_fail;
                     }
-                    break;
-                case 'k':
-                    if (out_keepdims != NULL && strcmp(str, "keepdims") == 0) {
-                        if (!PyBool_Check(value)) {
-                            PyErr_SetString(PyExc_TypeError,
-                                        "'keepdims' must be a boolean");
-                            goto fail;
-                        }
-                        *out_keepdims = (value == Py_True);
-                        bad_arg = 0;
+                }
+                else {
+                    /* Just copy reference */
+                    Py_INCREF(value);
+                    *(PyObject **)output = value;
+                }
+            }
+            else {
+#if PY_VERSION_HEX >= 0x03000000
+                PyErr_Format(PyExc_TypeError,
+                             "'%S' is an invalid keyword to ufunc '%s'",
```

Comment:
Wait, CPython3x format strings have a special syntax for unicode? I think there's a lot of places that cleanup could go